File: mobile/android/app/src/main/java/com/tigerwallet/app/trading/FuturesService.java
package com.tigerwallet.app.trading;

/**
 * Futures Trading Service - Android Implementation
 * Perpetual futures trading
 */

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FuturesService {
    
    public static class FuturesPair {
        public String id;
        public String base;
        public String quote;
        public String symbol;
        public double price;
        public double change24h;
        public double volume24h;
        public double high24h;
        public double low24h;
        public double makerFee;
        public double takerFee;
    }
    
    public static class FuturesPosition {
        public String id;
        public String userId;
        public String symbol;
        public String side;
        public double size;
        public double entryPrice;
        public double markPrice;
        public int leverage;
        public double margin;
        public double pnl;
        public double pnlPercent;
        public double liquidationPrice;
        public String marginMode;
        public long openTime;
    }
    
    public static class FuturesOrder {
        public String id;
        public String userId;
        public String symbol;
        public String side;
        public String type;
        public double size;
        public double price;
        public double filled;
        public String status;
        public int leverage;
        public String marginMode;
        public long createTime;
    }
    
    private static final String[] bases = {"BTC", "ETH", "BNB", "SOL", "XRP", "DOGE", "ADA", "AVAX", "DOT", "LINK", "MATIC", "LTC", "UNI", "ATOM", "XLM", "NEAR", "APT", "ARB", "OP", "INJ"};
    private static final double[] prices = {43250.0, 2280.0, 312.5, 98.75, 0.62, 0.082, 0.58, 38.2, 7.85, 14.50, 0.92, 72.30, 6.25, 10.45, 0.125, 3.25, 9.80, 1.12, 2.45, 35.50};
    
    private Map<String, Double> priceMap = new HashMap<>();
    
    public FuturesService() {
        initializePrices();
    }
    
    private void initializePrices() {
        for (int i = 0; i < bases.length; i++) {
            priceMap.put(bases[i], prices[i]);
        }
    }
    
    public List<FuturesPair> getPairs() {
        List<FuturesPair> pairs = new ArrayList<>();
        String[] quotes = {"USDT", "USDC"};
        
        for (int i = 0; i < bases.length; i++) {
            for (String quote : quotes) {
                if (!bases[i].equals(quote)) {
                    FuturesPair pair = new FuturesPair();
                    pair.id = "futures_" + i;
                    pair.base = bases[i];
                    pair.quote = quote;
                    pair.symbol = bases[i] + "/" + quote;
                    pair.price = prices[i];
                    pair.change24h = (Math.random() * 10 - 5);
                    pair.volume24h = prices[i] * 1000000;
                    pair.high24h = prices[i] * 1.05;
                    pair.low24h = prices[i] * 0.95;
                    pair.makerFee = 0.02;
                    pair.takerFee = 0.04;
                    pairs.add(pair);
                }
            }
        }
        return pairs;
    }
    
    public FuturesOrder openPosition(String userId, String symbol, String side, 
                                   double size, double price, int leverage, String marginMode) {
        FuturesOrder order = new FuturesOrder();
        order.id = "futures_order_" + System.currentTimeMillis();
        order.userId = userId;
        order.symbol = symbol;
        order.side = side;
        order.type = "MARKET";
        order.size = size;
        order.price = price;
        order.status = "PENDING";
        order.leverage = leverage;
        order.marginMode = marginMode;
        order.createTime = System.currentTimeMillis();
        return order;
    }
    
    public FuturesOrder placeOrder(String userId, String symbol, String side, String type,
                                  double size, double price, int leverage, String marginMode) {
        FuturesOrder order = new FuturesOrder();
        order.id = "futures_order_" + System.currentTimeMillis();
        order.userId = userId;
        order.symbol = symbol;
        order.side = side;
        order.type = type;
        order.size = size;
        order.price = price;
        order.status = "PENDING";
        order.leverage = leverage;
        order.marginMode = marginMode;
        order.createTime = System.currentTimeMillis();
        return order;
    }
    
    public void cancelOrder(FuturesOrder order) {
        order.status = "CANCELLED";
    }
    
    public double calculateLiquidationPrice(double entryPrice, int leverage, String side) {
        double liquidationPercent = 1.0 / leverage;
        if ("LONG".equals(side)) {
            return entryPrice * (1 - liquidationPercent);
        } else {
            return entryPrice * (1 + liquidationPercent);
        }
    }
    
    public double calculatePnL(double entryPrice, double closePrice, double size, String side) {
        if ("LONG".equals(side)) {
            return (closePrice - entryPrice) * size;
        } else {
            return (entryPrice - closePrice) * size;
        }
    }
}

File: mobile_apps/ios_legacy/TigerWallet/Services/Trading/FuturesService.swift
//
//  FuturesService.swift
//  TigerWallet
//
//  Futures Trading Service - iOS Implementation
//

import Foundation

struct FuturesPair {
    var id: String
    var base: String
    var quote: String
    var symbol: String
    var price: Double
    var change24h: Double
    var volume24h: Double
    var high24h: Double
    var low24h: Double
    var makerFee: Double
    var takerFee: Double
}

struct FuturesPosition {
    var id: String
    var userId: String
    var symbol: String
    var side: String
    var size: Double
    var entryPrice: Double
    var markPrice: Double
    var leverage: Int
    var margin: Double
    var pnl: Double
    var pnlPercent: Double
    var liquidationPrice: Double
    var marginMode: String
    var openTime: Date
}

struct FuturesOrder {
    var id: String
    var userId: String
    var symbol: String
    var side: String
    var type: String
    var size: Double
    var price: Double
    var filled: Double
    var status: String
    var leverage: Int
    var marginMode: String
    var createTime: Date
}

class FuturesService {
    
    private let bases = ["BTC", "ETH", "BNB", "SOL", "XRP", "DOGE", "ADA", "AVAX", "DOT", "LINK", "MATIC", "LTC", "UNI", "ATOM", "XLM", "NEAR", "APT", "ARB", "OP", "INJ"]
    private let prices: [String: Double] = [
        "BTC": 43250.0, "ETH": 2280.0, "BNB": 312.5, "SOL": 98.75, "XRP": 0.62,
        "DOGE": 0.082, "ADA": 0.58, "AVAX": 38.2, "DOT": 7.85, "LINK": 14.50,
        "MATIC": 0.92, "LTC": 72.30, "UNI": 6.25, "ATOM": 10.45, "XLM": 0.125,
        "NEAR": 3.25, "APT": 9.80, "ARB": 1.12, "OP": 2.45, "INJ": 35.50
    ]
    
    func getPairs() -> [FuturesPair] {
        var pairs: [FuturesPair] = []
        let quotes = ["USDT", "USDC"]
        
        for (index, base) in bases.enumerated() {
            for quote in quotes {
                guard base != quote else { continue }
                
                let price = prices[base] ?? 10.0
                let pair = FuturesPair(
                    id: "futures_\(index)",
                    base: base,
                    quote: quote,
                    symbol: "\(base)/\(quote)",
                    price: price,
                    change24h: Double.random(in: -5...5),
                    volume24h: price * 1000000,
                    high24h: price * 1.05,
                    low24h: price * 0.95,
                    makerFee: 0.02,
                    takerFee: 0.04
                )
                pairs.append(pair)
            }
        }
        return pairs
    }
    
    func openPosition(userId: String, symbol: String, side: String, size: Double, price: Double, leverage: Int, marginMode: String) -> FuturesOrder {
        return FuturesOrder(
            id: "futures_order_\(Int(Date().timeIntervalSince1970 * 1000))",
            userId: userId,
            symbol: symbol,
            side: side,
            type: "MARKET",
            size: size,
            price: price,
            filled: 0,
            status: "PENDING",
            leverage: leverage,
            marginMode: marginMode,
            createTime: Date()
        )
    }
    
    func placeOrder(userId: String, symbol: String, side: String, type: String, size: Double, price: Double, leverage: Int, marginMode: String) -> FuturesOrder {
        return FuturesOrder(
            id: "futures_order_\(Int(Date().timeIntervalSince1970 * 1000))",
            userId: userId,
            symbol: symbol,
            side: side,
            type: type,
            size: size,
            price: price,
            filled: 0,
            status: "PENDING",
            leverage: leverage,
            marginMode: marginMode,
            createTime: Date()
        )
    }
    
    func cancelOrder(order: inout FuturesOrder) {
        order.status = "CANCELLED"
    }
    
    func calculateLiquidationPrice(entryPrice: Double, leverage: Int, side: String) -> Double {
        let liquidationPercent = 1.0 / Double(leverage)
        if side == "LONG" {
            return entryPrice * (1 - liquidationPercent)
        } else {
            return entryPrice * (1 + liquidationPercent)
        }
    }
    
    func calculatePnL(entryPrice: Double, closePrice: Double, size: Double, side: String) -> Double {
        if side == "LONG" {
            return (closePrice - entryPrice) * size
        } else {
            return (entryPrice - closePrice) * size
        }
    }
}
